### PR TITLE
Refactor Cutadapt, add 3' mRNA-seq workflow

### PR DESCRIPTION
**Pull request by jkokosar**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/reads_processing/cutadapt.yml [hunk 1]
```diff
@@ -130,24 +145,28 @@
...
-       and not trailing
-       and not crop
-       and not headcrop
-       and not minlen %}
+      {% if not adapters.up_primers_file
+       and not adapters.down_primers_file
+       and not adapters.up_primers_seq
+       and not adapters.down_primers_seq
+       and not adapters.polya_tail
+       and not modify_reads.leading
```
Comment (mzganec) [inline resolwe_bio/processes/reads_processing/cutadapt.yml]: Consider adding `modify_reads.nextseq_trim` here.

File: resolwe_bio/processes/reads_processing/cutadapt.yml [hunk 2]
```diff
@@ -25,93 +21,112 @@
...
+    - name: modify_reads
+      label: Read trimming
+      group:
+        - name: nextseq_trim
+          label: NextSeq-specific quality trimming
+          type: basic:integer
+          required: false
+          description: |
+            NextSeq-specific quality trimming (each read). Trims also dark
+            cycles appearing as high-quality G bases. This option is mutually
```
Comment (mzganec) [inline resolwe_bio/processes/reads_processing/cutadapt.yml]: In one of the future versions we should make a drop-down menu for selection of quality trimming type. This would make mutual exclusion mentioned in the description more obvious.

File: resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    This 3' mRNA-Seq pipeline is comprised of QC, preprocessing,
+    alignment and quantification steps.
+
+    Reads are preprocessed by __Cutadapt__ which removes adapters, trims
+    reads for quality from the 3'-end, and discards reads that are too short
+    after trimming. Preprocessed reads are aligned by __STAR__
+    aligner. For read-count quantification, the __FeatureCounts__ tool
+    is used.
+
+    QC steps include downsampling, QoRTs QC analysis and alignment of
```
Comment (mzganec) [inline resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml]: Consider rephrasing the downsampling part. Downsampling does not contribute any information, i.e. report, to QC. I would therefore make a distinction from the rest of the tools. Consider using something along the lines _QC operates on downsampled reads and includes ..._

File: resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    - name: globin_reference
+      label: Indexed Globin reference sequence
+      type: data:genomeindex:star
+      description: |
+        Reference sequence index prepared by STAR aligner indexing tool.
+    - name: show_advanced
+      label: Show advanced parameters
+      type: basic:boolean
+      default: false
+    - name: downsampling
```
Comment (mzganec) [inline resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml]: Listing downsampling options before cutadapt options is misleading. It may lead the user into thinking that filtering and trimming is done on downsampled FASTQ files.

File: resolwe_bio/processes/reads_processing/cutadapt_3prime.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                            "cycles appearing as high-quality G bases. This option is mutually "
+                            "exclusive with the use of standard quality-cutoff trimming and is "
+                            "suitable for the use with data generated by the recent Illumina "
+                            "machines that utilize two-color chemistry to encode the four bases.",
+                default=10,
+            )
+
+            quality_cutoff = IntegerField(
+                label="Quality cutoff",
+                description="Trim low-quality bases from 3' ends of each read before adapter "
```
Comment (mzganec) [inline resolwe_bio/processes/reads_processing/cutadapt_3prime.py]: ends -> end

File: resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    - name: cutadapt
+      label: Cutadapt filtering
+      hidden: '!show_advanced'
+      group:
+      - name: quality_cutoff
+        label: Reads quality cutoff
+        type: basic:integer
+        required: false
+        description: |
+          Trim low-quality bases from 3' ends of each read before
```
Comment (mzganec) [inline resolwe_bio/processes/workflows/cutadapt_star_fc_quant_tf.yml]: ends -> end

File: resolwe_bio/processes/reads_processing/cutadapt_3prime.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            '-m', inputs.options.min_len,
+            '-O', inputs.options.min_overlap,
+            '-g', 'truseq=A{18}AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC',
+            '--discard-trimmed',
+            '-o', '{}_trimmed.fastq.gz'.format(name),
+            '-',
+        ]
+
+        # Run Cutadapt, write analysis reports into a report file
+        (Cmd['cutadapt'][first_pass_input] | Cmd['cutadapt'][second_pass_input] | Cmd['cutadapt'][third_pass_input]) \
```
Comment (mzganec) [inline resolwe_bio/processes/reads_processing/cutadapt_3prime.py]: Why don't we run cutadapt in multiple threads (`-j` option) here as well?
